test/model/recent_senders_test.dart

```diff
@@ -0,0 +1,153 @@
+import 'package:checks/checks.dart';
+import 'package:test/scaffolding.dart';
+import 'package:zulip/model/recent_senders.dart';
+import '../example_data.dart' as eg;
+
+void main() {
+  group('MessageIdTracker', () {
+    late MessageIdTracker idTracker;
+
+    void prepare() {
+      idTracker = MessageIdTracker();
+    }
+
+    test('starts with no ids', () {
+      prepare();
+      check(idTracker.debugIds).isEmpty();
+    });
+
+    test('calling add(id) adds the same id to the tracker just one time', () {
+      prepare();
+      check(idTracker.debugIds).isEmpty();
+      idTracker.add(1);
+      idTracker.add(1);
+      check(idTracker.debugIds.singleOrNull).equals(1);
+    });
+
+    test('ids are sorted ascendingly, with maxId pointing to the last element', () {
+      prepare();
+      check(idTracker.debugIds).isEmpty();
+      idTracker.add(1);
+      idTracker.add(9);
+      idTracker.add(0);
+      idTracker.add(5);
+      check(idTracker.debugIds).deepEquals([0, 1, 5, 9]);
+      check(idTracker.maxId).equals(idTracker.debugIds.last);
+    });
+  });
+
+  group('RecentSenders', () {
+    late RecentSenders recentSenders;
+
+    void prepare() {
+      recentSenders = RecentSenders();
+    }
+
+    test('starts with no stream or topic senders', () {
+      prepare();
+      check(recentSenders.debugIsEmpty).equals(true);
+    });
+
+    test('only processes a stream message', () {
+      prepare();
+      final dmMessage = eg.dmMessage(from: eg.selfUser, to: [eg.otherUser]);
+      recentSenders.handleMessage(dmMessage);
+      check(recentSenders.debugIsEmpty).equals(true);
+
+      final streamMessage = eg.streamMessage(
+        id: 100,
+        sender: eg.user(userId: 10),
+        stream: eg.stream(streamId: 1),
+        topic: 'topic',
+      );
+      recentSenders.handleMessage(streamMessage);
+
+      final expectedStreamSenders = {
+        1: {
+          10: MessageIdTracker.fromIds([100])
+        }
+      };
+      check(recentSenders.debugStreamSenders).deepEquals(expectedStreamSenders);
+
+      final expectedTopicSenders = {
+        1: {
+          'topic': {
+            10: MessageIdTracker.fromIds([100])
+          },
+        },
+      };
+      check(recentSenders.debugTopicSenders).deepEquals(expectedTopicSenders);
+    });
+
+    test('adding multiple messages', () {
+      prepare();
+      final message1 = eg.streamMessage(
+        stream: eg.stream(streamId: 1),
+        topic: 'topic1',
+        sender: eg.user(userId: 10),
+        id: 300,
+      );
```

Comment:
```suggestion
      final stream1 = eg.stream(streamId: 1);
      final user10 = eg.user(userId: 10);
      final message1 = eg.streamMessage(
        stream: stream1, topic: 'topic1', sender: user10, id: 300);
```

Then stream1 and user10 can be reused in the other messages.